include/deal.II/base/template_constraints.h

```diff
@@ -461,6 +464,63 @@ struct ProductType<unsigned int,float>
   typedef float type;
 };
 
+/**
+ * declaration of products for sacado types
+ */
+#ifdef DEAL_II_WITH_TRILINOS
+typedef Sacado::Fad::DFad<double> fad_double;
+typedef Sacado::Fad::DFad<fad_double> fad_fad_double;
```

Comment:
Let's not create these typedefs but instead use the fully qualified names below.